### ISSUE TITLE
Silent failure when replying with http status code not defined for endpoint

### ISSUE THREAD

**Issue by christopherraa** [jhthorsen/mojolicious-plugin-openapi#80]
I ran into a slight surprise today and might have happened upon a place where `M::P::OpenAPI` can be improved a little bit. My case was that I had an endpoint with a couple of responses defined (200, 202, 400, etc etc), but for some mysterious reason I got a `HTTP 500` error back from the endpoint, even when calling with 100% verifiable valid data.

It turns out that my error was that the endpoint tried to answer with `HTTP 201`, which was not defined in the `responses` block. This is absolutely an error on my end, but what ended up costing me quite a lot of time was the the OpenAPI-plugin responded with a `HTTP 500` and an error message that made absolutely no sense when looking at what was sent to the endpoint and what was done during processing. Here is an excerpt from the log:

```
[Mon Jun 25 21:15:52 2018] [debug] POST "/api/v1/foo" (58239db4)
[Mon Jun 25 21:15:52 2018] [debug] Routing to controller "My::Controller::Foo" and action "bar"
[Mon Jun 25 21:15:52 2018] [warn] OpenAPI >>> POST /api/v1/foo [{"message":"Missing property.","path":"\/errors"}]
[Mon Jun 25 21:15:52 2018] [debug] 500 Internal Server Error (0.001297s, 771.010/s)
```

When I removed the usage of the OpenAPI-plugin from the code and ran the test-suite against the API I got one test failing, telling me that the API returned a `HTTP 201` when a `HTTP 202` was expected. As you can see from the above log output there are absolutely no hints there as to the nature of the error, simply that _something_ has happened.

I agree that replying with something that is _outside_ what has been specified for the endpoint should result in a `HTTP 500`, but should / could the plugin perhaps do some warning in the log about the nature of the deviation from spec? In my opinion this is not something that should be exposed to the client, but something that very well could be logged at `fatal`-level in the application log.

The issue I am describing might very well be multiple separate issues. It might be summarized to the following qustions:

 - What is the expected behaviour when trying to respond with invalid status code for a particulat endpoint?
 - Should this not be logged so it can be followed up on?
 - Why does it respond with an error message saying that the property `/errors` is missing?

I'm open to brainstorming around what would be expected behaviour. I'm also not alien to the idea that I'm missing something here, so please correct me if this is covered in the docs and that I shoudl have seen what was the error at fault some other (obvious?) way.

**Comment by jhthorsen** [comment]
I understand that this is confusing, but it's actually by design: The plugin adds a "default" response to handle the default cases for status 400 and 500. This default response expects a response such as this:

```
{"errors":[...]}
```

So the log message `[warn] OpenAPI >>> POST /api/v1/foo ...` tells you that there's no `errors` property present in the 201 response that you created, which again results in "500 Internal server error" because your 201 response does not match the default error document.

Too see more how this works, you can expect the code around these lines:

* https://github.com/jhthorsen/mojolicious-plugin-openapi/blob/master/lib/Mojolicious/Plugin/OpenAPI.pm#L20
* https://github.com/jhthorsen/mojolicious-plugin-openapi/blob/master/lib/Mojolicious/Plugin/OpenAPI.pm#L102
* https://github.com/jhthorsen/json-validator/blob/master/lib/JSON/Validator/OpenAPI.pm#L113

It might have been a bad design from my side, where I chose to add "default" response instead of just "400" and "500", but it is more convenient for my usecases.

What I can do however, is to inform about falling back to "default" response. Something like this:

```
diff --git a/lib/JSON/Validator/OpenAPI.pm b/lib/JSON/Validator/OpenAPI.pm
index b265d89..59f1652 100644
--- a/lib/JSON/Validator/OpenAPI.pm
+++ b/lib/JSON/Validator/OpenAPI.pm
@@ -108,9 +108,14 @@ sub validate_request {

 sub validate_response {
   my ($self, $c, $schema, $status, $data) = @_;
+  my $blueprint = $schema->{responses}{$status};
   my @errors;

-  if (my $blueprint = $schema->{responses}{$status} || $schema->{responses}{default}) {
+  if (!$blueprint and $blueprint = $schema->{responses}{default}) {
+    $self->_log(debug => "Falling back to default response, since no schema is defined for status=$status");
+  }
+
+  if ($blueprint) {
     push @errors, $self->_validate_response_headers($c, $blueprint->{headers})
       if $blueprint->{headers};
```

What do you think?

**Comment by jhthorsen** [comment]
You can also look for "default_response" here: https://metacpan.org/pod/Mojolicious::Plugin::OpenAPI#register

**Comment by christopherraa** [comment]
That change you propose there would be a great step forward in my opinion. Had such an error message been present in the logs I would have seen exactly what kind of nit I had been. ;)

**Comment by jhthorsen** [comment]
It's not an error though. It's a design decision. That's why the message has `debug` and not `error` as log level. I'm not pushing the change yet. I would like more opinions on the subject first, since there's probably a better solution.

**Comment by jhthorsen** [comment]
Another idea could be to warn if `app->mode eq "development"`. That way it might be easier to spot, and I think the message might only be useful while developing.

**Comment by christopherraa** [comment]
I am not protesting that the server returns a 500. That part I agree with 100%. It is after all an endpoint to behaving according to spec. What I question is that no hints are given, and that the error-message provided is directly misleading. The change you proposed where the developer is given some feedback as to the nature of the error is in my opinion all that is needed. When the module _knows_ what is wrong then I see no reason to hide that information from the user of the module.

**Comment by christopherraa** [comment]
Absolutely, this is a feature you'd mostly be need in development, but I can see situations where it might happen in production as well (for those endpoints that are not under sufficient tests).

**Comment by sjn** [comment]
The pattern "I tried to do [something] with $this and $that, and failed" in error messages is a good one in any case.

I'd say apply your fix, perhaps also improve any other places in the code where error messages could benefit from a little context, and voilà the world is a teeny bit better. :)

**Comment by christopherraa** [comment]
Absolutely agree with @sjn here.

@jhthorsen Might be that I am missing something here, but how can the module _not_ know that you are trying to return a status code that is not in the spec? Does it not check against the spec and then throw an error? Or am I misunderstanding something?

**Comment by jamesbriggs** [comment]
Hi Jan.

There's 2 usability issues here that can be easily improved:

1) the debug message should say what response code it got and what it expected

2) 500 is not the best choice for OpenAPI response errors. 501 (not implemented) or 520 (Cloudflare empty or invalid response) allow monitoring software to distinguish between web server and application errors.

Returning 500 is a headache for Operations staff, and results in permanently muting that monitor.

FYI: I recently migrated some Perl applications from Swagger2 to OpenAPI, and it was brutal - one of the worst experiences I've ever had. Please make an effort to add some additional debugging messages and documentation.

Thanks, James.

**Comment by jhthorsen** [comment]
@jamesbriggs: Sorry, but I can't say I care much about the "operations headache". If your server keeps on throwing 500, then it should be fixed.

Everyone: #88 has implemented changes that hopefully will make it harder to come into this situation. Please see my comments on the PR. I don't think I will make a release until I get "go" from someone else.
